[#9255] Migrate instructor course edit page to Angular

joanneong opened the pull request:
Part of #9255

v7 look:

![image](https://user-images.githubusercontent.com/24493463/51343996-9bbfbf80-1ad2-11e9-8d78-31ad97ef3338.png)

v6 look:

![image](https://user-images.githubusercontent.com/24493463/51344004-a1b5a080-1ad2-11e9-99df-8f5e31d717fe.png)

Summary of what has been done:
- Migrated actions include getting course details, editing course details, deleting course, adding instructor, editing instructor, deleting instructor and sending reminder emails
  - Each action involve a new api endpoint, with the exception of deleting instructor which modified `DeleteInstructorAction`
- Migrated front-end ~(incomplete)~
- Migrated tests for most actions except for deleting instructor, which has to be written from scratch (it seems)

Some observations so far:
- Some visible lag time taken to load the page
- Timezones are missing the exact UTC offset as the `TimezoneService` does not seem to provide it yet
- ~Auto-detect button is not functional - again, `TimezoneService` does not seem to support this~
- Change in position for status messages in course edit page (used to be in between the course details panel and the instructor details panels)
- Code has not been refactored - html can be repetitive for instructors panel (can consider extracting into components in the future) - for Modals as well
- Horrible way of circumventing non-null checks since using `!` will fail `npm run lint`
- When you refresh the page, the instructor panels are reordered by google id in alphabetical order
- Because I was using reactive forms in the html for the page, I decided to avoid using template syntax as far as possible so two-way binding may appear extraordinarily lengthy. Not too sure if this is a good way to go about doing things

To do:
- [x] Write `DeleteInstructorActionTest`
- [x] Route links correctly (e.g. `View Details`)
- [x] Create modals for instructor privileges
- [x] Edit style for displaying google id for instructors when there is no id available yet
- [x] Add button to resend invitation email for a newly added instructor
- [ ] Check that no interactions have been missed out
- [ ] Delete old files after migration

wkurniawan07 commented:
> * Auto-detect button is not functional - again, `TimezoneService` does not seem to support this

Just use `moment` library? This is already done at least in admin sessions page.

joanneong commented:
I actually referenced that page at first to find the `TimezoneService`, bu

joanneong commented:
Progress update: I have missed out a panel for when the `Custom` role is selected. Other than that, the rest of the interactions should be completed (but will still need to check through all the files to confirm).

wkurniawan07 commented:
@joanneong you need to resolve the merge conflicts.

wkurniawan07 commented on src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts:
```diff
@@ -12,13 +71,627 @@ import { ActivatedRoute } from '@angular/router';
+  onSubmitAddInstructor(formAddInstructor: FormGroup): void {
+    // Create a copy of the added instructor
+    const addedInstructor: InstructorAttributes = {
+      googleId: formAddInstructor.controls.googleId.value,
```
This is causing bug that prevents instructor from being added. When adding a new instructor, they would not have a google ID yet.

wkurniawan07 commented on src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.html:
```diff
@@ -1,3 +1,368 @@
+                <div class="col-sm-9">
+                  <span class="form-control-static" id="role-view-{{ i }}">
+                    {{ instructorList[i].role }}
+                    <a class="view-role-details" href="#" (click)="viewInstructorRole(viewInstructorRoleModal, i)">
```
`href="#"` will bring you to the top of the page. You're likely looking for `href="javascript:;"`?

wkurniawan07 commented on src/web/app/pages-instructor/instructor-course-edit-page/instructor-course-edit-page.component.ts:
```diff
@@ -12,13 +71,627 @@ import { ActivatedRoute } from '@angular/router';
+  deleteCourse(): void {
+    const paramsMap: { [key: string]: string } = {
+      courseid: this.courseToEdit.id,
+      next: '/web/instructor/courses',
```
There shouldn't be a need to pass this to the back-end.

joanneong commented:
Updates:
- Being able to access a soft-deleted course is actually a bug carried over from v6 (you can access with the same link as well). I have fixed this by checking that the course is not soft-deleted before displaying the page. I have also added a test to test for this situation.
- Tooltip for the `Delete` button for an instructor panel is sometimes unstable (?)
- Still looking for a better solution to the DOM manipulation
- Still missing custom privileges
